Docs/StyleGuide/Style-Guide.md

```diff
@@ -0,0 +1,287 @@
+    // so that the method call is Integer.ToString(i)
+    // instead of Integer.IntegerToString(i).
+
+    // YES
+    method ToString(i: int) returns (s: string)
+        ...
+
+    // NO
+    method IntegerToString(i: int) returns (s: string)
+        ...
+}
+```
+
+## Code Layout
+
+### Braces
+Opening braces go on the same line by default.
+```dafny
+module M {
+    ...
+    method Met() {
+        ...
+    }
+}
+```
+In case the method (or function, lemma, etc) signature is too long to fit in one line, or in case the signature has at
+least one specification clause, the opening brace goes on a new line.
+
+```dafny
+module M {
+    ...
+    method Met(i: int) returns (j: int)
+        requires i % 2 == 0
+        ensures j > 10
+    {
+        ...
+    }
+}
+```
+
+This applies to every scope: `module`, `class`, `predicate`, `if`, `while`, and more.
+
+### Imports
+
+By default, import modules without opening them.
+```dafny
+import Coffee
+...
+```
+
+However, if some members of a module are used very frequently, import it using `opened`:
+```dafny
+import opened Donut
+...
+```
+When a file uses two modules, and both of them define a method of the same name, do not import them `opened`.
+```dafny
+import MyModule
+import YourModule
+...
+method MyMethod() {
+    MyModule.foo();
+    YourModule.foo();
+}
+```
+
+In this case, if you want to shorten their name, import them with a shorthand name.
+```dafny
+import M = MyModuleWithACumbersomeName
+import Y = YourModuleWithACumbersomeName
+...
+method MyMethod() {
+    M.foo();
+    Y.foo();
+}
+```
+
+Common imports, such as `StandardLibrary` and `Native`, should be grouped together, followed by custom module imports
+with a blank line in-between.
+
+```dafny
+import opened StandardLibrary
+import opened Native
+
+import opened Donut
+import Coffee
+```
+
+Although not required, it's recommended to keep the order of `import`s and `include`s alphabetically, except when it
+makes more sense to group them logically.
+
+## Indentation and Line Breaks
+
+### Tabs or Spaces?
+Spaces are preferred over tabs. Tabs should only be used to remain consistent with the existing code with tabs.
+
+Use 4 spaces for each indentation.
+
+### Maximum Character Limit
+Although there is no strict requirement, it is generally recommended to have a maximum of 120 characters per line.
+
+### Functions, Methods, Predicates, and Lemmas
+Every Dafny method has the following signature.
+```dafny
+method {:<attributes>} MethodName(param1: Type, param2: Type) returns (ret: Type)
+    decreases param1
+    requires P()
+    modifies param2
+    ensures Q()
+```
+
+When possible, put MethodName and the `returns` statement on the same line, as the keyword `returns` is distinct from
+other method specification clauses, such as `decreases`, `requires`, `modifies`, and `ensures`, which should appear in
+this order. Each method specification clause should be on a separate line.
```

Comment:
I tend to put the `decreases` clause last, because it's relevant only to mutually recursive functions/methods. In this way, `decreases` feels more like it's more tied to the function/method implementation than it is part of a "public specification".